samza-core/src/main/java/org/apache/samza/clustermanager/AbstractContainerAllocator.java

```diff
@@ -198,12 +202,20 @@ protected final SamzaResourceRequest peekPendingRequest() {
    * @param preferredHost name of the host that you prefer to run the processor on
    */
   public final void requestResource(String processorId, String preferredHost) {
-    SamzaResourceRequest request = getResourceRequest(processorId, preferredHost);
+    requestResourceWithDelay(processorId, preferredHost, Duration.ZERO);
+  }
+
+  public final void requestResourceWithDelay(String processorId, String preferredHost, Duration delay) {
+    SamzaResourceRequest request = getResourceRequestWithDelay(processorId, preferredHost, delay);
```

Comment:
Adding a delay to the timestamp on the request only delays the "matching/assignment" of the request to a resource. However the request will be sent to the RM immediately, since invoking 
`issueResourceRequest(request) `
will call 
`manager.requestResources(request)`
which will cause the AMRMClient to issue the request to the RM.

Now, at this instant, since the RM has not marked the NM-bearing-failed-container as unhealthy, the request will land at the RM and it may allocate a resource on the same NM again.
Whenenver this request is matched with this "bad" resource, it will fail again, and hopefully the second (or subsequent) retry of this request will be after the RM has marked the NM as unhealthy (due to exponential backoff). 

So either we should make the actual requests to the RM spaced out in time (exponentially), or at the very least document that the "matching" of requests to resource is spaced out exponentially.

Reply:
Ah thanks for the clarification. The logic makes more sense to me now. The documentation implies the requests are enqueued and not executed right away. The explanation of "issuing requests" vs "matching requests" makes the code more clear. I'll update the docs to reflect the true logic and I'll update the logic implementation to handle this.

Reply:
Will separate the the delayed requests from the current `requestsQueue`